### ISSUE TITLE
More concise syntax

### ISSUE THREAD

**Issue by elsurudo** [fotinakis/swagger-blocks#19]
I realize that swagger-blocks aims to provide a 1:1 mapping with the Swagger JSON spec, but is there any interest in having an (optional) more concise syntax (possibly similar to swagger-docs gem)?

It just doesn't sit right with me how low-density my swagger blocks specifications are, as they constitute a huge percentage of my controller classes. It makes working with the code a bit cumbersome.

I think this would be pretty simple to do... has anyone given this any thought?

**Comment by fotinakis** [comment]
I understand that for sure, it's definitely is a much longer-form format than swagger-docs. But one of the motivations in creating swagger-blocks in the first place was I found that swagger-docs's shorter, very positional-argument-heavy interface often needed it's own reference so I could figure out what argument position was what, only to also find out then that the arguments were incomplete in terms of covering the Swagger spec—probably because it's much more work in a gem like this to define a bespoke shorter interface and write the translation layer from that custom DSL to Swagger's DSL while retaining full compatibility.

So, even though I almost always think shorter code is better, honestly I don't think the costs of complicating the internal code and complicating user usage are worth it in this case, plus the cost of having to thoroughly document the new interface because it wouldn't just fall out of the Swagger spec itself—there are a lot of edge cases that might surprise you, and I want this gem to easily support all of them.

Some alternatives that you probably know: swagger-blocks doesn't care where you put the blocks, so if their length really gets in the way you could move them to another object, even one that is entirely just for documentation. I also find myself using the "outline view"/"jump to method name" feature of [your favorite editor here] more frequently in large controllers.

Closing for now, but if you'd like to keep the conversation going or suggest a particular concise syntax format I'm happy to hear more.

**Comment by elsurudo** [comment]
I understand where you are coming from, but it seems that something like:

```
parameter do
  key :name, :email
  key :in, :form_body
  key :description, "Email of user"
  key :required, true
  key :type, :string
end
```

can be represented as:
`parameter name: :email, in: :form_body, description: "Email of user", required: true, type: :string`

without loss of any generality.

Thoughts?

**Comment by elsurudo** [comment]
Basically, anywhere they `key` method is being used over and over should just have the option of accepting a hash instead. Seems like you're re-inventing the hash in a more verbose way.

**Comment by fotinakis** [comment]
That might work pretty well, actually.

I need to think more about how to handle cases like this though:

```ruby
parameter do
  key :name, :pet
  key :in, :body
  key :description, 'Pet to add to the store'
  key :required, true
  schema do
    key :required, [:name]
    property :id do
      key :type, :integer
      key :format, :int64
    end
  end
end
```

All the standard keys would probably be fine:
```ruby
parameter name: :pet, in: :body, required: true, description: 'Pet to add to the store'
```

But as soon as you want to include the schema, things get more complicated. I might be able to make this work:
```ruby
parameter(name: :pet, in: :body, required: true, description: 'Pet to add to the store') do
  schema(required: [:name]) do
    property(:id, type: :integer, format: :int64)
  end
end
```

...let me give this some more thought! Slightly crunched for time at the moment, but I think this is probably viable and wouldn't be too hard to implement. I'd be happy to review a PR from you if you have time to take a stab at this.

**Comment by elsurudo** [comment]
Yep, your suggestion is basically what I was going to suggest. I hope to take a look at this sometime next week. Basically anywhere where a `key` method is acceptable, we can merge all those key/value pairs with the hash passed to the parent method (`parameter`, in this case).

**Comment by fotinakis** [comment]
Cool, sounds good. The `key` method is accepted everywhere because it's in the base node internally, so it'd just be modifying that node to iterate over the options hash and add keys. The more complicated part is that we should pass `options` through every method everywhere, which might be slightly tedious because we currently pass version info around everywhere like `.call(version: version, &block)` for supporting both versions. This is probably a good opportunity to make those just `.call(options, &block)` and do some magic to merge options everywhere.

If you get time for this you can run the tests simply with `rspec` and add things to them [here](https://github.com/fotinakis/swagger-blocks/blob/master/spec/lib/swagger_v2_blocks_spec.rb) and [here](https://github.com/fotinakis/swagger-blocks/blob/master/spec/lib/swagger_blocks_spec.rb). I'd want to maintain test coverage for both syntaxes, but there's maybe a clear way of adding enough coverage of it to the current tests without forking the whole suite.

**Comment by sirwolfgang** [comment]
:+1:

**Comment by makeitnew** [comment]
I have updated the library to shorten the syntax as described in this thread and will submit a pull request sometime in the next couple of days.

**Comment by fotinakis** [comment]
Looking forward to reviewing it @makeitnew, thanks!

**Comment by mojidabckuu** [comment]
What is about usage of the same param list? I have methods that have the same params like create and update for 5 models. I think it is good idea to create refs for parameters lists.

**Comment by fotinakis** [comment]
Fixed via #39
